akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSupervisionSpec.scala

```diff
@@ -0,0 +1,656 @@
+/**
+ * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.fusing
+
+import scala.util.control.NoStackTrace
+import akka.stream.Supervision
+import akka.stream.stage.PushPullStage
+import akka.stream.stage.Context
+import akka.stream.stage.Directive
+import akka.stream.stage.Stage
+import akka.stream.stage.PushStage
+import akka.stream.stage.TerminationDirective
+
+object InterpreterSupervisionSpec {
+  val TE = new Exception("TEST") with NoStackTrace {
+    override def toString = "TE"
+  }
+
+  class RestartTestStage extends PushPullStage[Int, Int] {
+    var sum = 0
+    def onPush(elem: Int, ctx: Context[Int]): Directive = {
+      sum += elem
+      ctx.push(sum)
+    }
+
+    override def onPull(ctx: Context[Int]): Directive = {
+      ctx.pull()
+    }
+
+    override def decide(t: Throwable): Supervision.Directive = Supervision.Restart
+
+    override def restart(): Stage[Int, Int] = {
+      sum = 0
+      this
+    }
+  }
+
+  case class BufferingTestStage(decider: Supervision.Decider, absorbTermination: Boolean = false) extends PushPullStage[Int, Int] {
```

Comment:
the name buffering is misleading, a "normal" buffer is always a detached element since it allows upstream and downstream to progress independently when the buffer is neither empty or full.